WorkerTaskManager

kaisalmen opened the pull request:
**Last Update 2021-01-28**

The `TaskManager` allows to register a task expressed by an initialization and an execution function with an optional comRounting function to be run a web worker. It creates one to a maximum number of workers that can be used for execution. Multiple execution requests can be handled in parallel of the main task, If all workers are currently occupied the requested are enqueued and the returned promise is fulfilled once a worker becomes available again.

**What the implement status?**
- Functionality outlined in #18234 has been implemented with the exception of cost parameter. Some additional features are already available.
- Standard workers are implemented including Main-Execution fallback
- Dependency loading for standard workers is available
- Module workers are implemented, dependencies are declared regularly and therefore no extra functionality is required
- `exec` and `init` functions can be declared in module and then be packaged in standard worker if needed. I used this to define the worker code once and support both code paths
- A new example is supplied
  - It uses a dat.gui to configure the example. Let it run, stop and reset.
  - It allows to select all outlined worker types (standard workers with dependencies, module workers that declare dependencies if required and standard workers that are executed on main).
  - It also allows to set the maximum amount of workers, the overall executions count and the number of meshes to keep.
  - It is a potentially infinitely running example executing over and over the same workers, This has already proven to be very helpful in identifying memory holes and in verifying that CPU is utilized as expected.
- JSDoc has been completed and TypeScript definitions are generated for free from it, see https://github.com/mrdoob/three.js/pull/19650#issuecomment-654444532
it just utilizes the parser
- Code has been moved to `examples/jsm/taskmanager`
- Class notation eases creation of worker code in case of standard workers.
- Execution queue has been added to `WorkerTaskManager` 

**OBJLoader2**
- `OBJLoader2` and `OBJLoader2Parallel` have been removed from three.js with R125. I will ensure it will work with `WorkerTaskManager` and also use simplified worker transfer functions (for all interested see [here](https://github.com/kaisalmen/WWOBJLoader/tree/dev))

**TODO**
- [x] Solve concurrent init issues when the same `TaskManager` is passed to multiple instances of the same loader/class requesting init
- [x] Optimize and review execution queue in `enqueueForExecution` and `_kickExecutions`
- [x] **Ongoing** Complete documentation
  - [x] Example
  - [x] JSDoc for typescript defintions
- [x] **Ongoing** Create a wrapper for `OBJLoader` (standard and module version).
- [x] **Ongoing** Provide easy to use utility functions to transfer `BufferGeometry`, `Material` (meta-information) and `Mesh` bi-driectionally between Main and workers. All "transferables" shall be treated as such. This replaces old functions available with `OBJLoader2`

**Questions**
- Shall we add cost parameter later?
- Is the example too complex?

**Later**
- On init the maximum amount of possible workers for all tasks are created. This is time consuming. I could makes sense to make the init lazy (= create extra worker when needed until max. is reached), but this adds complexity to initialization especially when Transferables are involved as they need to be cached and cloned.

trusktr commented:
(There's a "Draft" feature if the PR is in progress. Useful for signaling to others it is WIP. Under the "Reviewers" section of the right sidebar, where it says "Still in progress? Convert to draft")

donmccurdy commented on examples/jsm/loaders/TaskManager.js:
```diff
@@ -0,0 +1,560 @@
+ * @author Kai Salmen / https://kaisalmen.de
+ */
+
+import { FileLoaderBufferAsync } from "./obj2/utils/FileLoaderBufferAsync.js";
```
Is this import from the `obj2` folder necessary?  It looks like it just adds a custom onProgress callback to the FileLoader, which may not be applicable to all implementations? If it is necessary it should probably be moved into the TaskManager file.

kaisalmen replied:
The necessity for it is gone. I created it for other prototyping activity and used it because it was there. With the introduction of `loadAsync` in `Loader` it collapsed to what it is now and I agree whatever is needed (using `FileLoader` with `setResponseType( 'arraybuffer'`) should be integrated directly in `TaskManager`.

kaisalmen replied:
Done

donmccurdy commented on examples/jsm/loaders/TaskManager.js:
```diff
@@ -0,0 +1,560 @@
+     * @param {String[]} [dependencyUrls]
+     * @return {TaskManager}
+     */
+    registerType ( type, maximumWorkerCount, initFunction, executeFunction, comRoutingFunction, dependencyUrls ) {
```
Are there benefits to having a separate maxWorkerCount for each type of task? The downside, I think, is that you can't effectively load balance across different types of task. If any worker can do any task, that balancing could be easier.

kaisalmen replied:
Maybe there is need for an overall count and a desired per `taskType` count. Was that your idea behind cost parameter?

> If any worker can do any task, that balancing could be easier.

Is this a realistic scenario?

donmccurdy replied:
I consider the `cost` parameter unrelated to whether or not there are separate worker queues for different task types.

My original idea had been to have a configurable number of workers for the entire TaskManager, say N. Each worker can perform any type of task, and all tasks are balanced across those N workers. So there would be no per-task worker limit, just the TaskManager total limit.

> Is this a realistic scenario?

Having multiple tasks running in parallel is realistic, at least. For example, when using both compressed geometry and compressed textures you'll need DRACOLoader and (one of) KTX2Loader or BasisTextureLoader. 

Whether there are very many realistic situations where shared queues are likely to give better performance? I'm not sure. If you think this is more complicated, or worse for some other reason, I'm also fine with leaving this as you have it for now.

kaisalmen replied:
> My original idea had been to have a configurable number of workers for the entire TaskManager, say N. Each worker can perform any type of task, and all tasks are balanced across those N workers. So there would be no per-task worker limit, just the TaskManager total limit.

Ok, now I see your point. Then we need to create N workers for every taskType either lazy or on init (cost are only extra memory and init time what becomes negligible if Workers really do intensive work). Then we can always use maximum of N for every taskType.
From my point of view the Worker should only include the code domain (init, exec) of one task as otherwise I fear mixing task and their dependencies into the worker will introduce unnecessary complexity and mixing module and non-module code will not be manageable. I hope this makes sense.

donmccurdy replied:
> Then we need to create N workers for every taskType... Then we can always use maximum of N for every taskType depending on current demand.

This is a bit different than what I meant. Something like...

```
var manager = new TaskManager().setWorkerLimit(2);
var objLoader = new OBJLoader(manager);
var ktx2Loader = new KTX2Loader(manager);
```

... would create only two Web Workers *total*, no matter how many loaders are using that manager, and no matter how many resources are loaded. OBJ and KTX2 tasks could be balanced across those two workers, and a single worker's queue would have tasks of both types.

There are some very real advantages to that approach: you have firm control over the _total_ number of Web Workers, and work is evenly distributed. But the concern of mixing module and non-module code makes sense. I'm not sure the module worker type is going to be worth a whole separate API in the long term, but I don't want to hold up this PR on the question – it's fine with me to continue as you have the PR for now.

kaisalmen replied:
> There are some very real advantages to that approach: you have firm control over the total number of Web Workers, and work is evenly distributed.

I agree and I like the simplicity of your approach. Unfortunately modules and dependencies make it very complicated to combine all code in one worker and then have n instances of them. I hope my understanding matches what you meant now.

What about the following compromise:
We follow the strict worker limit. The proposed `TaskManager` could be modified to internally create N worker for each registered loader. Taking your example this means 2*OBJLoader2 workers + 2*KTX2Loader workers are created. When tasks are add/executed it only runs two in parallel (2*OBJ+0*KTX, 1*OBJ+1*KTX, 0*OBJ + 2*KTX), the other spares are not utilized. Worker creation should then move to the *register* step. A *dispose* function could be added to the worker contract to signal runtime data clean-up.

donmccurdy replied:
> When tasks are add/executed it only runs two in parallel ...

This will be very complicated to enforce. What happens if we get requests in this order:

1. 10 OBJ parse tasks
2. 2 KTX2 parse tasks

After step (1) I would expect that the manager assigns ~5 OBJ tasks to each OBJ worker. But then when it gets to step (2) there are already two workers running, do we have to wait until one of the workers clears its queue before assigning KTX tasks to a KTX worker? What if more OBJ tasks arrive in the meantime?

I would prefer to avoid all that... maybe we have a `setWorkerLimit( n )` method on the TaskManager, and for now it will create N workers _per task type_ and use them all. Then in the future, if we decide to, we can try to combine the worker pool so that tasks share workers for more equitable load balancing.

kaisalmen replied:
Add task puts the processing requests to a fifo queue (currently Promises are stored in array). Whenever a worker signals exec complete a task is taken from the queue and a fitting worker is kicked. What is done in the current `TaskManager`/example is not so far off expect for the overall count is per task type and not globally. But getting there not be hard. I will prototype this.

donmccurdy replied:
I see! I wonder whether the time to signal the main thread that a task is done and then wait for a new task to come back has a measurable impact on completion times? Thinking of a model with 100+ Draco geometries for example... that doesn't seem like much of a problem though.

kaisalmen replied:
In lastest commit the maximum amount (N) of parallel worker execution is now handled on `TaskManager` level. N workers are created of each type (except for main fallback), but only the maximum defined in `TaskManager` are executed in parallel.

kaisalmen replied:
The previous code did not work as described. I pushed another update that really does works the way it does.
Init time of workers increases linearly with quantity even if Blobs/Strings are cached. For module workers there seems no obvious way to cache code except for pre-loading the module in the embedding html.

Execution on main is still strange. The `FakeTaskWorker` is now treated identically during execution, but when it solely used in the example than it completely blocks any rendering. When it is intermixed with real workers, then the rendering is not blocked. Manual triggering of requestAnimationFrame did not help. For now I leave it as it is

donmccurdy commented on examples/jsm/loaders/TaskManager.js:
```diff
@@ -0,0 +1,560 @@
+     * @param {string} type The type as string
+     * @return boolean
+     */
+    supportsType ( type ) {
```
Minor: I think I find the term "task" or "taskType" clearer than just "type" here, and for the methods below.

kaisalmen replied:
Yes, this is inconsistent.

kaisalmen replied:
Changed in latest commit.

donmccurdy commented on examples/jsm/loaders/TaskManager.js:
```diff
@@ -0,0 +1,560 @@
+     * @param {string} workerJsmUrl The URL to be used for the Worker. Module must provide logic to handle "init" and "execute" messages.
+     * @return {TaskManager}
+     */
+    registerTypeJsm ( type, maximumWorkerCount, workerJsmUrl ) {
```
I've never seen the term "jsm" anywhere outside the name of this folder, so I'd prefer to use a clearer term here. It's also important that even though the "examples/js" folder is being removed later this year, that has very little to do with whether this "jsm" method should be used — DRACOLoader, BasisTextureLoader, and KTX2Loader will all use the non-JSM register method. What do you think of:

```js
// Default.
manager.registerTask( type, maxCount, initFn, execFn, ... );

// URL. (undecided which name..)
manager.registerTaskModule( type, maxCount, moduleUrl );
manager.registerTaskUrl( ... );
manager.registerTaskSrc( ... );
```

I think the fact that the script happens to be an ES Module is less important than the fact that it's an external script. That means any user with a build system like webpack or browserify will need to specifically configure their application to output files in the right directories.

kaisalmen replied:
Agreed, let's remove jsm from method signatures. `module` is the best replacement term as it is also in line with the naming of the `Worker` constructor parameter, I think.

kaisalmen replied:
Changed in latest commit.

donmccurdy reviewed (commented):
Thanks, this is looking good! A few comments and questions.

donmccurdy commented:
> Should we adjust loaders in scope of this PR or do it later?

I think one loader in this PR is probably enough to get the general idea.

kaisalmen commented:
> I think one loader in this PR is probably enough to get the general idea.

Do you have any preference? My comfort zone is OBJ, but I am also aware that it is not the most requested thing. 😉 Should we collaborate? Is R119 a possible target?

kaisalmen commented:
Cool, you can use the TypeScript compiler to generate ts.d files from the jsdoc declarations.

kaisalmen commented:
@donmccurdy and @mrdoob I have fully updated the description. R120 is a realistic target, I think. Sorry it took so long. If I could do this work in my regular job, we would have been here month ago. I should be able to solve open issues in the upcoming weeks. So, if you don't identify potential blockers, then we could merge this before end of August!

kaisalmen commented:
@donmccurdy and @mrdoob I didn't make it for R120, but the `TaskManager` is now feature complete. `OBJLoader2Parallel` has been fully adapted to it. I stumbled upon some nasty async init problems while adapting it to the `TaskManager` especially when sharing the `TaskManager` in different loader instances. These are now under control.

Shall we rename the `TaskManager` to something else? The name is wired to the Windows TaskMananger in my head. What's your opinion?

donmccurdy commented:
Thanks @kaisalmen! I hope to review this in detail soon...

I also associate the name with windows task manager, but it seems like a pretty good name. I like the "Manager" suffix for its similarity to the existing LoadingManager. Maybe WorkerManager? Either that or TaskManager seem fine to me.

What do you mean by "legacy workers" in the PR description?

How confident are you in the updates to the various loaders that depend on this? I guess I'm wondering if we should try to get in the TaskManager itself first, as a clean PR with nothing depending on it, or merge this all at once.

kaisalmen commented:
>  Maybe WorkerManager? Either that or TaskManager seem fine to me.

Then let's keep the name `TaskManager`. The name is good and fits. It is just the association, but that's alright. 😄 

> What do you mean by "legacy workers" in the PR description?

Classic, normal or standard workers, so one can distinguish between `Module Workers` and let's say `Standard Workers` instead of just `Workers`. What do you think is the best term? `Standard Worker`? It's really a pitty `Module Workers` are not supported by Firefox or Safari. 

> How confident are you in the updates to the various loaders that depend on this?

I have removed the need `WorkerExecutionSupport` and accompanying code from `OBJLoader2Parallel` and I have verified that all examples work (also the additional examples/tests I have in my dev repo https://github.com/kaisalmen/WWOBJLoader/tree/TaskManagerPrototype). Apart from that no Loader has been touched, yet.
I also think that the new worker creation approach is better. Using the `class` notation simplifies the creation of strings (whole class or just single functions).
Separation would work, but I need to change the example and merge the `OBJLoader2` related changes afterwards. Take your time to look at the code and then we can discuss?!

I need to look again at the execution loop of the `TaskManager`. It behaves like a thread that sleeps for a short while. I had created an endless loop and that is the current state of the solution, but it may not be good enough as it slows down processing, but seems robust at least.

donmccurdy commented:
I'm worried about giving the impression that Module Workers are the better/recommended option, and using the term "legacy" to describe the alternative gives that impression. The key difference between the two is not the use of ES Modules, but the fact that Module Workers rely on external scripts with a pre-determined URL. That's going to be very hard to use with modern bundlers without requiring users to individually put their dependencies in the right place.

Note that we are in the process of deprecating `examples/js` in favor of ES Modules in `examples/jsm`, which has no particular relation to Module Workers. I expect that the loaders I'll update to use TaskManager (BasisTextureLoader, DRACOLoader, and KTX2Loader) will be ES Modules themselves, but will not use Module Workers. I'm worried that the Legacy/Module naming will be pretty confusing for users in that process.

kaisalmen commented:
It was not my intention to make a recommendation here. I have already adjusted the wording to `standard` in the description. Honestly, `Module Workers` can't even be preferred as only Chromium-based browsers are able to use them. `TaskManager` is not biased here as you can load dependencies and add extra code pieces for standard workers. Standard and module workers have feature parity here. 

>  but the fact that Module Workers rely on external scripts with a pre-determined URL. That's going to be very hard to use with modern bundlers without requiring users to individually put their dependencies in the right place.

But, isn't that even more problematic when you used importScripts` in standard workers? You can use the same module in main and module workers without problems (see `OBJ2LoaderWorker` usage in `OBJLoader2Parallel` for creation of both worker types https://github.com/kaisalmen/three.js/blob/TaskManagerProto/examples/jsm/loaders/OBJLoader2Parallel.js#L119-L130). Logically, I don't see why bundlers should not be able to handle this. But I need to acquire more knowledge regarding the js bundler black magic.


What I esthetically like about module workers that they remove the inconsistency when you use modules in the main and you have to use a different syntax in the worker (`importScripts` vs. module syntax). It may also encourage one to write things only once. 😄

kaisalmen commented:
Finally, the PR is ready for review. 🚀 I am still updating the documentation of the new example, but this does not block the review.

mrdoob commented:
How about naming this `WebWorkerManager`?

kaisalmen commented:
... or `WorkerTaskManager` because tasks are executed by workers? The am fine with any of these names.

mrdoob commented:
That works too 👌

mrdoob commented on examples/files.js:
```diff
@@ -113,6 +113,7 @@ var files = {
 		"webgl_loader_prwm",
 		"webgl_loader_stl",
 		"webgl_loader_svg",
+		"webgl_loader_workertaskmanager",
```
No longer in alphabetical order 🤓

kaisalmen replied:
It was late. 😉

kaisalmen replied:
Corrected, rebased and ast three commits are squashed

kaisalmen commented on examples/jsm/loaders/OBJLoader2Parallel.d.ts:
```diff
@@ -1,23 +1,19 @@
+    private _buildWorkerCode;
+    _executeWorkerParse(content: any): void;
 }
+import { OBJLoader2 } from "./OBJLoader2.js";
```
Because I was lazy and now use the typescript compiler to generate the d.ts files from jsdoc (see https://github.com/mrdoob/three.js/pull/19650#issuecomment-654444532 ).

mrdoob replied:
I see... And jsdoc can't be configured?

kaisalmen replied:
The function param types defined in jsdoc are used by `tsc` to specify the types in `d.ts`. I overlooked that the import and export statements now contain the `.js` file extension. This is wrong. I either need to manually correct this or find a way to configure `tsc` to automatically create the correct output (preferred).

kaisalmen committed:
Squashed all commits due to remvoal of OBJLoader2.

WorkerTaskManager:
- Remove need for execution loop requiring wait
- Rename TaskManager To WorkerTaskManager and move it to directory WorkerTaskManager below loaders
- Is now able to properly handle multiple requests for initTaskType. If one is ongoing the other requests are not answered until the first one is complete. This prevents that parsing starts in any sequentially triggered loaders before init is really complete.
- All context (legacy, module or main) use TaskManagerDefaultRouting.comRouting if no other routing function is provided
- Introduced execution loop
- Dependency loading allows to provide strings along with to be loaded urls
- Adjusted naming (taskType and module)
- Reorganize execution flow and renamed addTask to enqueueForExecution
- Align FakeWorker main execution behaviuor with TaskWorkers: The maximum no of workers is created for all types
- Moved maximumWorkerCount from WorkerTypeDefinition to TaskManager. Maximum number of workers for every type are created, but only maximum number is executed.
- Removed the need for FileLoaderBufferAsync
- Fixed init bugs due to broken async await
- Added default comRouting function
- Added TaskWorker (extension of Worker) and FakeTaskWorker (used on main)
- Added jsm worker where mesh is created on main and is then shipped to worker which uses buffers as base for new meshes

webgl_loader_taskmanager.html
- OBJLoader replaces OBJLoader2 in example
- now features mtl loading for female and male obj models
- uses code from existing modules for export to legacy
- Added Reset Button that can reset the selected parameters and the complete scene without reload
- Allow to define overall execution, loop count and meshes to keep visible
- Adjusted naming (taskType and module)
- Add gui controls which allow selection of workers and quantity of instances + abort
- Added fake worker that runs on main and not in worker. Prepare code for future UI selection
- jsm workers rely on imported comRouting
- Add a second jsm based worker to the example
- Add and remove meshes no longer synched with renderer as otherwise sphere don't get cleaned on tab change (no rendering)
- One million executions with stable resource usage

TransferableUtils:
- split into TransferableUtils and MeshMessageStructure to have a now have a formal definition of what should is transferred.

General:
- Transformed all other relevant files to classes and fixed broken examples
- Updated typescript definitions
- Completed documentation and typescript definition files. TypeScript compiler is used to generate d.ts files from jsdoc descriptions
- default comRouting function (jsm or legacy case) takes init and exec functions as input

- Make init and excute fully async
- All workers in experiment example load three in Worker
- Experiment with transferable content generated in Worker

- Created TaskManager and WorkerDefinition derived from proposal of Don McCurdy #18234
- addTask returns Promises for every added task, but enqueues them when maximum count is reached
- example adds 1000 task, worker loops from 1 to 10^8 on every worker execution, 8 parallel workers are available

kaisalmen commented:
@mrdoob and @donmccurdy I have squashed and then rebased the branch. I replaced `OBJLoader2` and `OBJLoader2Parrallel` with a worker wrapper around `OBJLoader` and got it working as both standard and module worker. There are  still some issue, but the overall concept works.

I still have to update the issue text and there left a note on top.

Quick ideas for way forward:
- Make transfer of Meshes and Materials between Main and Workers more accessible and understandable to the users (= evolution and condensation of `TransderableUtils`, `MeshReceiver` and `MaterialHandler` ), because this is the key for adapting `WorkerTaskManager` by others, I think. I realised this when I saw https://github.com/mrdoob/three.js/pull/21035 .
- Polish example and then we could merge it in.

Interesting observation:
- When the ES6 class notation is used, then it can be easily put to a string. This is very useful when you want to create standard workers from ES6+ code which uses imports and exports. When you don't use the class notation this tasks is very error-prone because the variant ways you can define an object.

Btw, the potential change to `OBJLoader` could be applied independently. I made `OBJLoader2` independent of `MTLLoader` a longer time ago.

kaisalmen commented:
> Make transfer of Meshes and Materials between Main and Workers more accessible and understandable to the users

I guess this was to vague. So I hope I can clarify this further: Instead of building extra logic/structures like I tried to achieve with `MeshReceiver` and `MaterialHandler` we should have tools that could "transmit" `BufferGeometry`, `Materials` and other required things between Main and worker. Some like `sendBufferGeometry`, `sendMaterial` or `sendMesh` which ensure all `Transferables` are handled as such and the data is handled as effectively as possible. On the receiving side the reverse functions need to exist. These is not directly required by `WorkerTaskManager` to work (it is currently not dependent on the older classes as well only the example code is), but it is needed to use Workers effectively`and as consequence adopt the `WorkerTaskManager` . Especially when you don't want to rewrite an existing loader or some other thing that you want to move to a worker.

Do these thought make sense? Who else could contribute to this discussion and give or discuss ideas?

donmccurdy commented:
> If this works and we get better building blocks for moving "things" bi-directionally between workers and Main, should we adapt another loader before merging this or shall we do this afterward. What are your thoughts?

The use cases I'm most interested in do not need to transfer complex full scenes, just textures (BasisTextureLoader, KTX2Loader) or a single geometry (DRACOLoader). So any extra building blocks mostly feel like overkill for me, but I understand if your use cases need a bit more.

kaisalmen commented:
Ok, the utility to ship `BufferGeometry` is basically there, but needs to be polished. This is needed to make the example work and will be a handy utility for others, I think. Everything else left over from `OBJLoader2` will be condense or removed as it is not required in the context of the `WorkerTaskLoader`. We could then discuss to finally merge the PR and think about further utilities later if required.

kaisalmen commented:
I found a nice way to "serialize" and "de-serialize" `BufferGeometry` where everything that can be a Transferable is handled like one. This is WIP and is currently in the external repo (https://github.com/kaisalmen/WWOBJLoader/tree/TransferableMeshes)

The testbed is this: https://github.com/kaisalmen/WWOBJLoader/blob/TransferableMeshes/examples/wtm_transferables.html
You can edit the code and run the example in gitpod if you are interested.

kaisalmen commented:
I have updated the description to reflect the removal of `OBJLoader2` and updated the TODO list.

The rework of the transfer utilities is ongoing in the external repo. I will push an update once the example works again including the `OBJLoader` wrapper and `OBJLoader2` is based on the new utils as well.

kaisalmen commented:
Good news: I ported `OBJLoader2`/`OBJLoader2Parallel` on the new [TransferableUtils](https://github.com/kaisalmen/WWOBJLoader/blob/TransferableMeshes/examples/jsm/loaders/workerTaskManager/utils/TransferableUtils.js) (working title). I need to fix all the examples now, but all functionality is there and it is more condensed and easier to use. Once the `WorkerTaskManager` is working again I will push the code.

Shipping a `BufferGeometry` in a worker:
```javascript
const bufferGeometry = new TorusKnotBufferGeometry( 20, 3, config.params.segments, config.params.segments );
new GeometryTransport( config.params.name, config.id )
	.setGeometry( bufferGeometry, 2 )
	.package( false )
	.postMessage( context );
```
... and unpacking it on Main is as easy as this.
```javascript
const mesh = new THREE.Mesh(
    new GeometryTransport().loadData( payload ).reconstruct( false ).getBufferGeometry(),
    new THREE.MeshPhongMaterial( { color: new THREE.Color( 0xff0000 ) } )
);
```
Right now there is support for these: `DataTransport`, `GeometryTransport`, `MeshTransport`, `MaterialsTransport` and additional util functions around Material storage and code serializations.

donmccurdy commented:
What might usage look like in a loader to transfer arbitrary data, e.g. a JSON object and two ArrayBuffers?

The structure of WorkerTaskManager is looking good here! I haven't reviewed things related to Material and Mesh transport since I'm not really familiar with those loaders.

kaisalmen commented:
This is the current [code]( https://github.com/kaisalmen/WWOBJLoader/blob/TransferableMeshes/examples/jsm/loaders/workerTaskManager/utils/TransferableUtils.js#L123-L178) and this is how can be used:

```javascript
new DataTransport() // `cmd` and `id` are optional, btw
        .setParams( jsonData ) // jsonData is be stored in params
	.addBuffer( 'modelData', arrayBufferData )
	.addBuffer( 'textureData', arrayBufferTextureData )
	.package( false ) // the boolean is there to be able to clone the transferable data. Important for shared init data
	.postMessage( context );
```

The base class provides you with `getMain()` and `getTransferasbles()`, so you the data to `WorkerTaskManager`. It does no depend on the utilities! And on both side you can transfer it back to the classes (`package`/`reconstruct` functions).

donmccurdy commented:
Ok! This sounds like it should work. :)

kaisalmen committed:
webgl_loader_workertaskmanager.html is now only using TransferableUtils in addition to WorkerTaskManager. All other utilities have been removed or transformed into TransferableUtils.

TransferableUtils:
- Provide simple means to package and reconstruct mesh related data bi-directionally
- Contains: TransportBase, DataTransport, GeometryTransport, MaterialsTransport, MaterialUtils, MaterialCloneInstruction, MeshTransport, CodeUtils, ObjectManipulator)
- Allow arraybuffer copy if needed (e.g init or geometry re-usage)
- Add MaterialStore: Perform all required tasks MaterialHandler did and remove the overhead.

WorkerTaskManager:
- Fixed transferables being transferred as such

Removed MeshReceiver and MaterialHandler.

kaisalmen commented:
The example is fully working again. It fully utilizes `TransferableUtils.js`, and I repeat the `WorkerTaskManager` has no dependency to it. 😄  Theses functions make building wrappers around existing loaders easier (see `tmOBJLoader.js` whtich supports standard and module loaders)

kaisalmen committed:
Merged TransportBase and DataTransport (inheritance: DataTransport -> MaterialsTransport and DataTransport -> GeometryTransport -> MeshTransport)
Aligned usage in tmOBJLoader.js and webgl_loader_workertaskmanager.html

kaisalmen committed:
Move MaterialUtils and MaterialStore to their own files. Renamed TransferableUtils to TransportUtils.
Updated Typescript definitions.

kaisalmen committed:
Move all worker to a single folder below "workerTaskManager".
Update TransportUtils documentation.

kaisalmen committed:
Updated documentation
Updated Typescript definitions
Removed MaterialCloneInstruction

kaisalmen commented:
@donmccurdy and @mrdoob This is now ready for review and from my point of view in really good shape now. I hope you share that view.
The only things that still needs correction is the description inside the example. I won't finish this tonight, but likely tomorrow.

kaisalmen committed:
Updated documentation of webgl_loader_workertaskmanager.html

kaisalmen commented:
Example documentation update is ready as well.

mrdoob commented:
Is there a way to reduce the `webgl_loader_workertaskmanager` example to 150-300 lines of code total?
Complicated examples are hard to maintain and users find them overwhelming... 😕

kaisalmen commented:
Mmmhh, not without removing the potentially, indefinitely running feature and the UI configuration. It could come down to run **n** `WorkerTaskManager` **m** times with one or two specific workers without UI options (=load&run).

What about:
- The mega-example stays as is in the external repo (there it is even bigger as it contains `OBJLoader2` as well)
- **Here**, I reduce the example as suggested an I should be able to land in the suggested length target
- I am committed to maintain this stuff

Are you considering to have an example bootstrap which other example could use? So, one does not need to do the standard scene setup or can really overwrite it when required? Wasn't there a discussion about this a longer time ago?

mrdoob commented:
> Are you considering to have an example bootstrap which other examples could use? So, one does not need to do the standard scene setup or can really overwrite it when required? Wasn't there a discussion about this a longer time ago?

I'm not a fan of that approach...

Instead, what I try to do is to make API changes that make the examples shorter. Like designing `THREE.InstancedMesh` so people can easily take advantage of `ANGLE_instanced_arrays` and with an [example that is 140 lines of code](https://github.com/mrdoob/three.js/blob/master/examples/webgl_instancing_dynamic.html).

kaisalmen commented:
> I'm not a fan of that approach...

I remembered that correctly then. 😉

Is my suggested approach ok? You don't see this in Wednesday's release, do you?

kaisalmen committed:
Simplified example

kaisalmen committed:
Removed unneeded workers

kaisalmen commented:
I have simplified the example. It is now 350 lines and contains one inline worker definition (30 lines). Shrinking it further will be hard.

mrdoob commented:
Looking much better! 👍
But yes, won't be able to review this properly before Wednesday, unfortunately.

kaisalmen committed:
Use fake function call in OBJLoader worker wrapper allowing it to work with the unaltered loader.

kaisalmen committed:
Fix missing EventDispatcher

kaisalmen commented:
I have identified a potential problem introduced by a recent change (last commit) in `BufferAttribute` . It's a one line change. Without the change a `BufferAttribute` cannot directly be sent via `postMessage`. I will open an independent PR for that.

I am currently moving all wtm related things to its own repo and make it completely independent of obj2. Then I can release npm packages for both. Switched to snowpack for dev + rollup for bundling. 👍

kaisalmen commented:
See #21769 and #21770 .

kaisalmen committed:
Align code with latest code from three-wtm (1.0.0-beta.6)

kaisalmen commented:
I just pushed an update. Code is now based on:
https://www.npmjs.com/package/three-wtm/v/1.0.0-beta.6

I want to release version three-wtm 1.0.0 and wwobjloader 4.0.0 until the end of this week. That is realistic, because I already fixed multiple bugs, especially init issues and I ensured that workers can created from uglified code. This was a pain point with OBJLoader2 in the past and it needed adjustments in both libraries. 🎉

kaisalmen commented:
@mrdoob and  @donmccurdy this is ready. The npm module is ready. I will release 1.0.0 after I have overhauled the README. I don't see any code change before release and therefore no additional changes from my side here.

mrdoob commented:
@donmccurdy I'll let you decide what to do with this 🤔

kaisalmen commented:
I have just released verion 1.0.0:
https://www.npmjs.com/package/three-wtm

kaisalmen commented:
@donmccurdy and @mrdoob What do we do? Are still interested with this? Is there anything I can do to make this any more desirable? Btw, the code has jsdoc comments that are used to generate d.ts files in the external repo. 😉

donmccurdy commented:
Sorry for the long delay here —

I'm struggling with what role WorkerTaskManager should play, and where it should be (examples/jsm, separate repo, etc.) to fulfill that. My original motivation for the TaskManager proposal in https://github.com/mrdoob/three.js/issues/18234 was to consolidate worker management logic required by KTX2Loader and DRACOLoader, and [WorkerPool.js](https://github.com/mrdoob/three.js/blob/dev/examples/jsm/utils/WorkerPool.js) has basically solved that need, with a very small amount of code.

Because this PR provides features like transferring Mesh, Material, and BufferGeometry instances over a wire, it's also about 20x larger. The serialization is valuable in its own right — like a better/faster `toJSON/fromJSON` — but wouldn't _ideally_ be built into a Web Worker framework, since you might also want that feature for Web Sockets, Node.js, databases, etc. The combined package is a good fit for loading OBJ (and perhaps other complex formats like COLLADA, STEP, or USD?) but those same formats tend to be complex enough to merit their own native libraries and GitHub repositories...

Do we think it is likely that this framework is also a good fit for something like @gkjohnson's three-mesh-bvh, or doing physics in a worker with cannon-es? I'm not convinced that it is, unfortunately. In comparison, consider a combination of...

- [greenlet](https://github.com/developit/greenlet) or [threads.js](https://threads.js.org/), and
- https://github.com/mrdoob/three.js/pull/21035 

... the combination is more flexible for use cases like physics, and provides a threading implementation compatible with both Node.js and Web Workers. I'm wondering if that is the way we should go here instead.

gkjohnson commented:
If code for transferring core types between web workers is going to be added I still strongly vote for the approach posed in #21035 or at least something that's usable separately from any worker infrastructure three.js is choosing to provide. The benefits of such an approach are not limited to use in transferring web workers, as well. Keeping geometry data as the original typed attributes (or even just cloning a typed array directly) is _much_ faster and the typed array can be stored in browser caches like IndexedDB and sent over websockets if needed.

kaisalmen commented:
> and WorkerPool.js has basically solved that need, with a very small amount of code.

Yes, I guess your intention what should be achieved with `WorkerTaskManager` and what I interpreted were fairly different. Me involving `OBJLoader2` from the beginning may have not been the best choice, because it brought many more requirements (support for different worker types, dependencies and no need for bundlers, etc.) to the table. Now, with it removed from three.js also the current need for the extra feature in the scope of three.js is gone.

> Because this PR provides features like transferring Mesh, Material, and BufferGeometry ... but wouldn't ideally be built into a Web Worker framework, since you might also want that feature for Web Sockets, Node.js, databases, etc.

Yes, I completely agree with you both. It is contained in this PR, but `WorkerTaskManager` is not dependent: It's just a handy tool that I needed again mostly to make `OBJ2Loader` work efficiently and make the example simpler. And, yes, if these transport or send-efficient-over-the-wire tools existed independently of any web worker stuff (e.g. #21035 is realized) that would be great.

> ... the combination is more flexible for use cases like physics, and provides a threading implementation compatible with both Node.js and Web Workers. I'm wondering if that is the way we should go here instead.

I actually thought about if using [comLink](https://github.com/GoogleChromeLabs/comlink) would make sense. Currently, `WorkerTaskManager` is independent from other projects apart from three.js (and this dependency is not strong). This is good, but it may nowadays no longer be the best choice.

[three-wtm](https://github.com/kaisalmen/three-wtm), like [OBJLoader2](https://github.com/kaisalmen/WWOBJLoader) exist as an independent projects now.  **three-wtm** could evolve into something completely independent of three.js (with transfer tools moved to OBJ2 or no longer required). I have to think about where I want to go with it and what makes sense.

I have the feeling that this PR won't get merged because it does not bring the right value to three,js, but if you think any of the code/work done here is useful in another scope (elsewhere in three.js or beyond), then let me know/let's talk. 🙂

kaisalmen commented:
Dear @mrdoob @donmccurdy and @gkjohnson I will now close this PR. I have waited this long, because I wanted the external project [three-wtm](https://github.com/kaisalmen/three-wtm). to reach a good state, first. Now is the time.

The code has been transformed to TypeScript and I performed clean-up, fixed bugs and added new features (e.g. different worker count for each registered task, generic `WorkerTask` isolated from `WorkerTaskDirector` (new name)). I have isolated a core library without any dependencies, so the `WorkerTaskDirector` can be used in other contexts and I created an extension for three.js:
https://www.npmjs.com/package/wtd-core/v/2.0.0-beta.1
https://www.npmjs.com/package/wtd-three-ext/v/2.0.0-beta.1

The code and documentation needs some polishing, but V2.0.0 will be there early May (won't have much time to work on in the upcoming week). Maybe this work can be of to you use even if the connection to three.js is getting weaker. Feedback from you is very welcome. If you like drop by the repo. I will announce the release on Twitter.

Btw, [OBJLoader2(Parallel)](https://github.com/kaisalmen/WWOBJLoader) is still alive and is now using it as well:
https://www.npmjs.com/package/wwobjloader2/v/5.0.0-beta.1

kaisalmen commented:
Release is almost ready. You can try the new examples here: http://wtd.kaisalmen.de/
